### ISSUE TITLE
[v1.3.4-evolve] indexed cross-entropy target returns wrong loss instead of rejecting

### ISSUE THREAD

**Issue by Gabriel-Kahen** [tsotchke/eshkol#552]
Observed with clean `tsotchke/eshkol@90cbd7130f47b8184bcc77b8d5c1b0026da980de`, `Eshkol Compiler v1.3.4-evolve`, compiler SHA-256 `caa295b19a6e9388963aa0def99dade63656d2dcbffccad421bd1daaa1db3750` on Linux x86-64. Upstream HEAD was not rebuilt.

```scheme
(define logits (reshape #(2.0 0.0 0.0 2.0) 2 2))
(define indices #(0 1))
(display (cross-entropy-loss logits indices))
```

AOT and JIT both exit 0 and print `2.8200751916029176`. For indexed targets `[0,1]`, the mean softmax cross-entropy for these logits is approximately `0.1269280110429726`.

If `cross-entropy-loss` supports indexed class targets, the returned value is wrong. If it accepts only one-hot targets, the rank/shape mismatch should be rejected explicitly rather than interpreted as another layout.

**Comment by tsotchke** [comment]
Confirmed on a current build — the wrong value reproduces exactly:

```
(display (cross-entropy-loss (reshape #(2.0 0.0 0.0 2.0) 2 2) #(0 1)))
2.8200751916029176
```

against the ~`0.1269280110429726` that indexed targets `[0, 1]` should give for those logits.

The way you framed it is the part that makes it actionable: either indexed class targets are supported and the value is wrong, or only one-hot targets are supported and a rank/shape mismatch must be rejected explicitly. What must not happen is the third thing, which is what happens today — the mismatch is silently reinterpreted as another layout and a plausible-looking number comes back. A loss that is wrong by a factor of twenty-two but finite and stable will not fail any smoke test; it will just train the wrong model.

**Status.** #558 addresses this and declares `Fixes #552` — cross-entropy validating its targets and computing the documented loss for both the indexed and one-hot forms, which is the "do both" resolution rather than picking one. It is open and currently **conflicting** with master; it needs a rebase before it can land. That is on us.
